File: leetcode_75/Binary_Tree_DFS/python/437.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    targetSum = 0

    def get_sum(self, node: TreeNode, s: int):
        if node is None:
            return 0

        s += node.val
        return int(s == self.targetSum) + self.get_sum(node.left, s) + self.get_sum(node.right, s)

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.targetSum = targetSum

        def wrapper(node: TreeNode):

            if node is None:
                return 0

            return self.get_sum(node, 0) + wrapper(node.left) + wrapper(node.right)

        return wrapper(root)

File: leetcode_75/Binary_Tree_DFS/cpp/437.cpp

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int pathSum(TreeNode *root, int targetSum)
    {
        this->targetSum = targetSum;

        return this->helper(root);
    }

private:
    int targetSum;

    int get_sum(TreeNode *node, long int s)
    {
        if (node == nullptr)
        {
            return 0;
        }

        s += node->val;
        return (s == this->targetSum) + this->get_sum(node->left, s) + this->get_sum(node->right, s);
    }

    int helper(TreeNode *node)
    {
        if (node == nullptr)
        {
            return 0;
        }
        return this->get_sum(node, 0) + this->helper(node->left) + this->helper(node->right);
    }
};